Failed MMS redownload

jezek opened the pull request:
Notify user when a MMS arrived but can't be read and add the ability to download it afterward.

For the long story: 
https://forums.ubports.com/topic/5100/the-mms-lost-story/110


Add redownload button to failed MMS bubble
- Redownload button visibility and enabled depends on status
- Error message text depends on button visibilty
- Some error details are in message text encoded in json (Description, Expiry, Size)

Prerequisites: Folowing PR's need to be merged to make this work ubports/nuntium#8, ubports/telepathy-ofono#20, [gitlab/ubports/core/history-service#8](https://gitlab.com/ubports/core/history-service/-/merge_requests/8) and ubports/telephony-service#20

fixes https://github.com/ubports/messaging-app/issues/17
fixes https://github.com/ubports/ubuntu-touch/issues/1408
fixes https://github.com/ubports/ubuntu-touch/issues/239

jezek committed:
Handle MMS error messages

Add redownload button to failed MMS bubble
Handle simple MMS message when cellular data is off
Get rid of warning in logs when displaying an error message

lduboeuf commented:
Awesome, will test it as soon as i can 👍 :)
I guess message will need to be changed. 
Is there any work to be done on history service to replace that empty MMS with the correct one ?.

jezek commented:
@lduboeuf 
> Is there any work to be done on history service to replace that empty MMS with the correct one ?.

Just edited the main comment, there are 2 PR's accompanying this. I tested only with nuntium-inject-push, it worked. But then made little changes before pushed to github, but it should work. Will test better & write some report tomorrow.

lduboeuf commented:
yessss, it works! ;-) don't know why it does not replace the error message automatically though, need to go back and return to conversation to get it refreshed and same for the Thread list. Maybe the model in history db doesn't trigger a changed signal for that case. I know it does for any changes in Status of the message. Maybe using the "MessageStatusPending" for that kind of message would be good when storing the "empty" mms first ? https://github.com/ubports/history-service/blob/xenial/src/types.h#L69, or we can create a new one like "MessageStatusDeferred", or of course we still force the List to change, but not a very nice solution

lduboeuf commented:
or well not sure for the status change, maybe just delete the message after retry action ? ( eventModel.removeEvents )

edit: or don't know if setting status directly in messaging-app when clicking "retry" to MessageStatusPending with eventModel.writeEvents(...) will do the trick. Tell me if you need help for the messaging-app part

jezek commented:
@lduboeuf Yes, I know about the quirks. This is just the first (somehow) working version. There is lot of work.

Until now, I have traced back the message mark-read signal from message-app to nuntium and used the pathway to send the re-download signal.

My next step is to trace the massage-add signal from nuntium to message-app and figure out, why other custom statuses I try to set in nuntium ("failed", "deferred") are not propagated. After I know more, changes will be made.

> or well not sure for the status change, maybe just delete the message after retry action ?
> or don't know if setting status directly in messaging-app when clicking "retry" to MessageStatusPending with eventModel.writeEvents(...) will do the trick.

Messaging-app should not make status changes. And until I figure out why are the other custom statuses ignored a proper solution is not possible. 



> Tell me if you need help for the messaging-app part

Yes, I will, but first I have to know with what you can help.

lduboeuf commented:
> My next step is to trace the massage-add signal from nuntium to message-app and figure out, why other custom statuses I try to set in nuntium ("failed", "deferred") are not propagated. After I know more, changes will be made.

Yes, there is a filter here, that only hande mms with "recieved" status: https://github.com/ubports/telepathy-ofono/blob/xenial/connection.cpp#L429

lduboeuf commented:
for the history-service, received message are managed here: https://github.com/ubports/history-service/blob/xenial/daemon/historydaemon.cpp#L1023

And for the notification part, here: https://github.com/ubports/telephony-service/blob/1d25bdea1eace05cacc8c0b3a73ad1129ce57cb9/indicator/textchannelobserver.cpp#L450

jezek commented:
Why there is an error on last check (e633218)?
The armhf build says:
```
...
The following tests FAILED:
	  1 - SingleSim (Failed)
	  3 - MessagesView (Failed)
...
```
and the arm64 build says:
```
...
The following tests FAILED:
	  1 - SingleSim (Failed)
	  2 - MessageBubble (Failed)
	  3 - MessagesView (Failed)
	  4 - AttachmentsDelegate (Failed)
	  5 - StickersHistoryModel (Failed)
...
```

But I haven't touched these files.

Is it cause of qt5.12 transition?

lduboeuf commented:
Yes, you should wait a couple of days, there is work on CI and Qt5.12 transition

lduboeuf commented on src/qml/MessagesListView.qml:
```diff
@@ -90,8 +90,8 @@ MultipleSelectionListView {
             var sourceFile = ""
-            var isMmsError = model.textMessageType === HistoryEventModel.MessageTypeMultiPart && model.textMessageAttachments.length === 0 && model.textMessage.length === 0
-            if (isMmsError) {
+            var isIncomingMmsError = model.senderId !== "self" && model.textMessageType === HistoryThreadModel.MessageTypeMultiPart && model.textMessageAttachments.length === 0 && (model.textMessage.length === 0 || model.textMessageStatus === HistoryThreadModel.MessageStatusTemporarilyFailed || model.textMessageStatus === HistoryThreadModel.MessageStatusPermanentlyFailed || model.textMessageStatu ...
```
@jezek so we don't need anymore the empty mesage check if i follow well ? (model.textMessage.length === 0)

jezek replied:
I would leave it there. So it will work on possible empty messages, which were meant as error.
If I used only the status check, all previous error messages, introduced by you, were not recognized as error messages.
The empty message, should be handled as permanent error state, which can't be redownloaded.

jezek commented:
@lduboeuf why did you add sticker feature into this failed MMS PR? Shouldn't it be a stand-alone PR? Or is it more convenient to add to this PR and test together?

lduboeuf commented:
@jezek  oh, did i made something wrong ?, i just synced with xenial

jezek commented:
@lduboeuf aaah, you merged with xenial. I don't know if it is wrong, I was told (sometimes) to do only PR commits, that are in the scope of the PR. So I learned to rebase, after the main branch changes (I know, rebase is a bit cumbersome and some are against it). But you're the maintainer here, your calls. I was just curious, why the new commits.
Note: I see you just merged using one commit. Feels much cleaner.

lduboeuf commented:
@jezek that should be ok now ? Hope i did not erase anything...

jezek commented:
@lduboeuf Seems ok. But for better sleep, today after I update telephony-service PR's notification bubble text, I will check this PR if everything is ok.

jezek commented:
@lduboeuf I saw that there is only one commit in this PR and I'm sure I've made more commits, so I was shocked when I saw only one. I thought you managed to accidentally managed to delete them. I was relieved, when I checked and found out, you only squashed them into one. So I would propose to also change the commit message, cause "Add redownload button to failed MMS bubble" is not the only thing, that was done in this PR. I would recommend "Handle MMS error messages"

lduboeuf commented:
@jezek i agree yes, better rename it, you can do ?

jezek commented:
@lduboeuf 
> i agree yes, better rename it, you can do ?

Done

lduboeuf commented:
I bet nuntium need update too... https://paste.ubuntu.com/p/g84pnfvP4G/

jezek commented:
@lduboeuf 
> I bet nuntium need update too...

I really don't know whats going on with the ```ubports-qa``` scripts. I compile & deploy ```nuntium``` manually (all other PRs using crossbuilder) and for me everything works (on Stable channel). The ```nuntium``` was already on the top of laest xenial branch. I added and deleted an "empty" commit to force checks again. I hope it helps.

lduboeuf commented:
I don't understand either. 
EDIT: now back to working state

fredldotme commented on src/qml/MessageAlertBubble.qml:
```diff
@@ -101,4 +170,8 @@ ListItemWithActions{
         onTriggered: eventModel.removeEvents([messageData.properties]);
     }
 
+    Component.onCompleted: {
```
I don't think we should leak SMS messages into the log file.
It could easily trip off security-conscious folks.

fredldotme replied:
I've been told that this component is only loaded when an error occurs, in which case I'm fine with logging it.

fredldotme commented:
Just nitpicking about visibility of messages in logs, other than that looks good!

lduboeuf commented:
> Just nitpicking about visibility of messages in logs, other than that looks good!

humm, the console.log is intentional when there is an error message, to give information for debuggin g, it should not be sensitive info

jezek commented:
@fredldotme 
> Just nitpicking about visibility of messages in logs, other than that looks good!

Like @lduboeuf said the log is in error message bubble and since it is an error message, there is no personal info. Its output looks like this:
```bash
phablet@ubuntu-phablet:~$ zcat .cache/upstart/application-legacy-messaging-app-.log.*|grep textMessage:
qml: textMessage: {"Code":"x-ubports-nuntium-mms-error-activate-context","Message":"forced debug error: error-activate-context","Expire":"2021-12-07T16:03:35+01:00","Size":29696,"MobileData":false}
qml: textMessage: {"Code":"x-ubports-nuntium-mms-error-activate-context","Message":"forced debug error: error-activate-context","Expire":"2021-12-07T16:19:47+01:00","Size":29696,"MobileData":false}
```
I think it is OK, but it was there only for debugging purposes and the piece of code can be removed. Shall I remove it?

lduboeuf commented:
> I think it is OK, but it was there only for debugging purposes and the piece of code can be removed. Shall I remove it?

IMO no, as it might be interesting for debugging if any issue

fredldotme commented:
Somehow I can't resolve the issue so I will mention it here again:

Thanks for the explanation, that makes sense and I'm fine with it.

Flohack74 commented on src/qml/MessageAlertBubble.qml:
```diff
@@ -27,21 +28,70 @@ ListItemWithActions{
-    height: errorTxt.height + textTimestamp.height + units.gu(2)
+    readonly property var messageStatus: messageData.textMessageStatus
+
+    readonly property string permanentErrorText: i18n.tr("Oops, there has been an error with the MMS system and this message could not be retrieved. Please ensure Cellular Data is ON and MMS settings are correct, then ask the sender to try again.")
```
Please make sure that translation updates also contain the changes to the pot file for Weblate to pick it up.

lduboeuf committed:
update translation file

lduboeuf committed:
Merge branch 'xenial' into xenial_-_failed-MMS-redownload
